### PR TITLE
Yaml Generation

### PR DESCRIPTION
**Pull request by FrenchBen**
**- What I did**
I added support for YAML output for the engine CLI

**- How I did it**
Extended on the Cobra docs output but for Yaml - Some of the changes have been pushed back to upstream: https://github.com/spf13/cobra/pull/380

**- How to verify it**
`make yamlfiles` will output Yaml files within the `yaml/yaml1` directory

**- Description for the changelog**
Added support for YAML file generation of Docker Engine CLI commands

**- A picture of a cute animal (not mandatory but encouraged)**

Bear it all

### CODE REVIEW

**Comment by aaronlehmann** [comment]
> make manpages will output Yaml files within the yaml/ directory

man pages are troff formatted, so it's a little strange that `make manpages` would output YAML. Is this a precursor format that later gets converted to man pages?

**Comment by FrenchBen** [comment]
@aaronlehmann bleh.. it looks like I missed some the changes to the repo structure. 
You're correct, this generates more than just manpages as I piggy backed on the function used in manpage.

The yaml output is used to build the docs.docker.com pages, but will never be stored within `docker/docker` - The idea is that our docs reflect 100% the cli code logic without having developers update the yaml files separately.

**Comment by justincormack** [comment]
This should be built in CI, otherwise no idea when it breaks.

**Comment by FrenchBen** [comment]
@justincormack that's the idea - the CI piece still needs to be added.

**Comment by LK4D4** [comment]
design is ok for me

**Comment by thaJeztah** [comment]
Can we make sure that the files in `man/src` are brought in sync with the docs/reference/commandline documentation? I think there may be information present in the docs/ files that are not present in the man pages (and vice-versa).

Also, what's the approach to link from the reference docs to other parts of the documentation? For example, https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md refers to various other parts of the docs, but the https://github.com/docker/docker/blob/master/man/docker-run.1.md man page (which is not in `man/src`) does not have those links (as man pages usually link to other man-pages, but not to online docs).

Our man pages have historically been a lot less complete than the online docs, so we should look carefully when moving to using those as documentation, as they are far from complete.

Do man-pages support tables? (see https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md#restart-policies---restart)

**Comment by tiborvass** [comment]
Ping @FrenchBen let us know if you need help with this.

**Comment by FrenchBen** [comment]
After conversation with @tiborvass and @thaJeztah the following will happen:

1. `docs/reference/commandline` will be the new source of truth for the extended descriptions
2. The above MD will need to partially remove the current CLI output within the markdown
3. `docs/reference/commandline` will need to be update to have all canonical commands and their aliases

**Comment by tiborvass** [comment]
@FrenchBen there are some compile errors due to unvendored dependency.

**Comment by thaJeztah** [comment]
We need to check license compatibility as well (https://github.com/jteeuwen/go-bindata);

```
This work is subject to the CC0 1.0 Universal (CC0 1.0) Public Domain Dedication
license. Its contents can be found at:
http://creativecommons.org/publicdomain/zero/1.0
```

**Comment by justincormack** [comment]
CC0 is fine, its as close to public domain as allowed.

**Comment by thaJeztah** [comment]
Thanks for checking @justincormack

File: yaml/Dockerfile
Comment (tiborvass) [inline yaml/Dockerfile]: you dont need this line.

**Comment by mstanleyjones** [comment]
Right now, the `yaml/Dockerfile` doesn't work. It needs to install `mercurial` and also there is some problem with where the `generate.sh` gets put vs the `WORKDIR`. I tried a few permutations but I couldn't figure it out.

Also, #30804 is now in `master`. It standardizes the format of the CLI references so that they will be easier for this PR to consume:

The long description should be grabbed from between `## Description` to `## Examples`. If there is no `## Description` then there is no long description.

The examples should be grabbed from between `## Examples` to the next H2-level heading, or to the end of the file, whichever comes first.

Both of these should be grabbed as a string of Markdown.

**Comment by vieux** [comment]
ping @FrenchBen can you take a look at @mstanleyjones's comments ?

**Comment by thaJeztah** [comment]
@FrenchBen told me during lunch that he was working on this :smile:

**Comment by vieux** [comment]
LGTM ping @thaJeztah

**Comment by FrenchBen** [comment]
Latest changes made by @tiborvass LGTM

File: yaml/Dockerfile
Comment (vieux) [inline yaml/Dockerfile]: why not use the `golang:1.7.5-alpine` ? /cc @tiborvass @FrenchBen

**Comment by tiborvass** [comment]
@FrenchBen I now see `exec: "gcc": executable file not found in $PATH` when I do `make yamlfiles`

File: yaml/Dockerfile
Comment (tiborvass) [inline yaml/Dockerfile]: I had that at some point. Will reevaluate but right now having problems with some dependencies requiring `gcc`

Comment (FrenchBen) [inline yaml/Dockerfile]: @tiborvass didn't we have gcc in the Dockerfile that you removed?

**Comment by tiborvass** [comment]
@FrenchBen @vieux updated

File: yaml/Dockerfile
Comment (thaJeztah) [inline yaml/Dockerfile]: whitespace looks a bit wiggly in this Dockerfile 😇

**Comment by vieux** [comment]
LGTM

File: yaml/vendor.conf
Comment (thaJeztah) [inline yaml/vendor.conf]: I may be confused, but we're vendoring docker/docker here, and create a custom client / generator using the _vendored_ code; as a result, the output of the commands / flags will never automatically update to match the current version of the source code?

Comment (thaJeztah) [inline yaml/vendor.conf]: Also, vendoring is done during build, so effectively, we're re-fetching all dependencies every time we `make yamlfiles`. Would it make sense to have the source code in a separate repository, so that we actually _vendor_ the packages in that repository?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Left some questions

File: yaml/Dockerfile
Comment (vieux) [inline yaml/Dockerfile]: indent look wrong

**Comment by tiborvass** [comment]
Will have to rethink this a little bit.

**Comment by FrenchBen** [comment]
giphy

**Comment by thaJeztah** [comment]
@frenchben no fear, we have a plan!

**Comment by tiborvass** [comment]
@thaJeztah @FrenchBen I updated this PR, squashed the commits, moved the `yaml/` folder to `docs/yaml/` and set up a `hooks` directory to be picked up by Docker Cloud. Currently the contents of that are for debugging.

After talking with @johndmulhausen we might need to send a PR (in a post_build hook) to docker/docker.github.io with the generated yaml files.

**Comment by tiborvass** [comment]
For people who want to test (@FrenchBen @thaJeztah @johndmulhausen) , I set up an autobuild pipeline on Docker cloud and here's the resulting image with only the yaml files:
```bash
$ docker pull tiborvass/docker:17.04.0-dev-docs
$ cid=$(docker create tiborvass/docker:17.04.0-dev-docs)
$ docker cp $cid:/docs /tmp/
$ docker rm $cid
$ cd /tmp/docs/
```

**Comment by thaJeztah** [comment]
@tiborvass after this is merged on master, we should probably get it into the release branch(es), so that the YAML / image can be automatically build for specific versions

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by mstanleyjones** [comment]
This iteration is missing the extended description and examples, which should be pulled from `## Description` and `## Examples` sections of the CLI reference Markdown files.

**Review by FrenchBen — COMMENTED** [review]
LGTM although I'm not familiar with the cloud pre-hooks, so can't speak of that. Overall YAML stuff is 👍

**Comment by FrenchBen** [comment]
I would like to see a Makefile update prior to merge, so that we can also build this manually and verify that future changes generate the proper YAML

File: docs/yaml/generate.go
Comment (FrenchBen) [inline docs/yaml/generate.go]: All of my previous changes to this logic was lost during re-write.

The folder structure is incorrect, as it looks for something like:
`commandline/container/stats.md`
when it should be looking for:
`commandline/container_stats.md`

**Review by FrenchBen — COMMENTED** [review]
🙅‍♂️  for the Source file parsing

File: docs/yaml/generate.go
Comment (tiborvass) [inline docs/yaml/generate.go]: You should still have it in your branch then. I only pushed what was already in this PR and squashed and rebased them without conflict.

Comment (FrenchBen) [inline docs/yaml/generate.go]: Possible fix:
```
func loadLongDescription(cmd *cobra.Command, path ...string) error {
        for _, cmd := range cmd.Commands() {
                if cmd.Name() == "" {
                        continue
                }
                fullpath := filepath.Join(path[0], strings.Join(append(path[1:], cmd.Name()), "_")+".md")

                if cmd.HasSubCommands() {
                        loadLongDescription(cmd, path[0], cmd.Name())
                }

                if _, err := os.Stat(fullpath); err != nil {
                        log.Printf("WARN: %s does not exist, skipping\n", fullpath)
                        continue
                }

                content, err := ioutil.ReadFile(fullpath)
                if err != nil {
                        return err
                }
                description, examples := parseMDContent(string(content))
                cmd.Long = description
                cmd.Example = examples
        }
        return nil
}
```

File: Makefile
Comment (FrenchBen) [inline Makefile]: 🎉

**Review by FrenchBen — COMMENTED** [review]
LGTM

File: Makefile
Comment (cpuguy83) [inline Makefile]: How about `yamldoc`?

**Comment by cpuguy83** [comment]
I don't know what this is all about, anyone else able to review?

**Comment by thaJeztah** [comment]
I think we're awaiting confirmation that these files are working correctly for the docs team @frenchben @tiborvass?

**Comment by mstanleyjones** [comment]
Yes they are not working yet. I'm working with Tibor on it.

**Comment by tiborvass** [comment]
I'll close it for now, as we cannot merge it. Will reopen once it's ready.

**Comment by mstanleyjones** [comment]
Latest output LGTM. Note that we need to be somewhat rigorous about the formatting of the CLI reference files. If there is a `Description` or `Examples` section, it needs to be a H2 section. And if there are any complicated tables with block level elements they need to be HTML tables. There may be more "gotchas" in future.

**Review by mdlinville — APPROVED** [review]

**Comment by tiborvass** [comment]
Ping @estesp @duglin not sure what powerpc did there: https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/361/console

File: Makefile
Comment (dnephin) [inline Makefile]: Something like `yaml-docs-gen` would be great (much closer to the name of the bundle)

**Review by dnephin — APPROVED** [review]
LGTM

Just a minor issue with the makefile target name

**Comment by duglin** [comment]
@tiborvass seems like the power issue was a flake, seems to have vanished.

**Review by justincormack — APPROVED** [review]

**Pull request merged**
